### PR TITLE
Make tide use HeadRef to get SHA instead of last commit if possible.

### LINKED ISSUE #6343: tide stuck because it's supplying the wrong head sha to merge
Tide seems to be stuck right now trying to merge #6205:
```
{"client":"github","level":"info","msg":"Merge(kubernetes, test-infra, 6205, {  c08a5adde90424a8dd52f5a04dcf206ed1e4399b merge})","time":"2018-01-19T00:17:26Z"}
{"component":"tide","error":"Head branch was modified. Review and try the merge again.","level":"info","msg":"Merge failed: PR was modified.","time":"2018-01-19T00:17:27Z"}
{"component":"tide","level":"info","msg":"Action: MERGE, Targets: [{6205 {fejta} {master refs/heads/} {test-infra kubernetes/test-infra {kubernetes}} {[{{{[{cla/linuxfoundation fejta authorized SUCCESS} {tide In tide pool. SUCCESS}]} c08a5adde90424a8dd52f5a04dcf206ed1e4399b}}]}}]","time":"2018-01-19T00:17:27Z"}
```

Looking at that PR, there are two commits:
```
f2759a5ca480cc54a5debe474aeaa7944f9cb8a1 (HEAD -> pr/6205) Add :branchprotector-image and :push targets.
c08a5adde90424a8dd52f5a04dcf206ed1e4399b Add bazelbuild/rules_k8s to WORKSPACE
```

Chronologically, c08a5adde90424a8dd52f5a04dcf206ed1e4399b is the latest commit, but f2759a5ca480cc54a5debe474aeaa7944f9cb8a1 is the tip of @fejta's branch.

For whatever reason, prow is supplying the wrong sha, so the merge is failing, and it just keeps retrying this PR, rather than moving on.

/assign @cjwagner 
cc @spxtr 
/area prow

### PR DESCRIPTION
**Pull request by cjwagner**
This is a partial revert of https://github.com/kubernetes/test-infra/pull/6225 

Currently Tide gets the latest commit sha for a PR by getting the 'last' commit via graphql. Previously it got the sha from `HeadRef.Target.OID` field, but we cannot read this field if the forked repo has been deleted so we switched to using the last commit. Unfortunately the 'last' commit is determined by author date not commit date so if commits are reordered non-chronologically tide tests the wrong commit and fails to merge the PR. 

This temporary fix uses the HeadRef if it exists and falls back to the potentially mis-ordered commits if it doesn't. This should only fail on PRs with mis-ordered commits where the forked repo has been deleted.

fixes #6343 
/area prow
/kind bug
/cc @rmmh @ixdy

### CODE REVIEW

File: prow/tide/tide.go [hunk 1]
```diff
@@ -685,3 +690,10 @@ type searchQuery struct {
 		}
 	} `graphql:"search(type: ISSUE, first: 100, after: $searchCursor, query: $query)"`
 }
+
+func (pr *PullRequest) SHA() string {
```
Comment (rmmh) [inline prow/tide/tide.go]: can you put the commit message in as a comment here? there's a lot of subtle edge cases you're handling here!

Comment (cjwagner) [reply prow/tide/tide.go]: Done.

**Comment by fejta** [comment]
This is deployed, right?

**Comment by cjwagner** [comment]
It is now!

**Comment by cjwagner** [comment]
This can be done properly now with the `headRefOid` field :tada: :tada: :tada: 
https://platform.github.community/t/some-prs-are-missing-head-refs/4586

I'll make a PR.

**Pull request merged**